[clang-tidy] Only expand <inttypes.h> macros in modernize-use-std-format/print

mikecrowe opened the pull request:
Expanding all macros in the printf/absl::StrFormat format string before conversion could easily break code if those macros are expanded change their definition between builds. It's important for this check to expand the <inttypes.h> PRI macros though, so let's ensure that the presence of any other macros in the format string causes the check to emit a warning and not perform any conversion.

mikecrowe committed:
[clang-tidy] Only expand <inttypes.h> macros in modernize-use-std-format/print

Expanding all macros in the printf/absl::StrFormat format string before
conversion could easily break code if those macros are expended to
change their definition between builds. It's important for this check to
expand the <inttypes.h> PRI macros though, so let's ensure that the
presence of any other macros in the format string causes the check to
emit a warning and not perform any conversion.

mikecrowe commented:
I'm not particularly happy with the `FormatStringConverter` constructor now taking seven parameters. The risk of them getting muddled isn't huge though because they are all of different types. Should I perhaps put them all into the `FormatStringConverter::Configuration` struct or perhaps a separate struct in the absence of named parameters? Even if I leave them as seven parameters, is there a preferred order for them? Any recommendations gratefully received.

I considered adding a configuration option to list other macros that should be permitted in the format string. I will investigate doing so if this change is acceptable.

OpenBSD and mingw32 seem to just use `PRI` prefixes in their `inttypes.h`. I don't know whether Microsoft and MacOS do anything like glibc does which required the special `__PRI` handling.

EugeneZelenko commented on clang-tools-extra/clang-tidy/utils/FormatStringConverter.cpp:
```diff
@@ -208,11 +208,22 @@ FormatStringConverter::FormatStringConverter(ASTContext *ContextIn,
     return;
   }
+
+  if (const auto MaybeMacroName =
```
Please do not use `auto` unless type is explicitly stated in same statement or iterator.

mikecrowe commented:
It looks like the tests are failing on Windows because the `%s` in the command line
```
// RUN: %check_clang_tidy \
// RUN:   -std=c++20 %s modernize-use-std-format %t -- \
// RUN:   -config="{CheckOptions: {StrictMode: true}}" \
// RUN:   -- -isystem %clang_tidy_headers \
// RUN:      -DPRI_CMDLINE_MACRO="\"%s\"" \
// RUN:      -D__PRI_CMDLINE_MACRO="\"%s\""
```
is being replaced by the filename for the LIT test:
```
| Running ['clang-tidy', 'C:\\ws\\src\\build\\tools\\clang\\tools\\extra\\test\\clang-tidy\\checkers\\modernize\\Output\\use-std-format.cpp.tmp.cpp', '-fix', '--checks=-*,modernize-use-std-format', '-config={CheckOptions: {StrictMode: true}}', '--', '-isystem', 'C:/ws/src/clang-tools-extra/test/../test\\clang-tidy\\checkers\\Inputs\\Headers', '-DPRI_CMDLINE_MACRO="C:\\ws\\src\\clang-tools-extra\\test\\clang-tidy\\checkers\\modernize\\use-std-format.cpp"', '-D__PRI_CMDLINE_MACRO="C:\\ws\\src\\clang-tools-extra\\test\\clang-tidy\\checkers\\modernize\\use-std-format.cpp"', '-std=c++20', '-nostdinc++']...
```
I can believe that this is being done by `TestRunner.py` (or some other part of llvm-lit) since that's expected for the first `%s` on the command line, but the surprise is that it doesn't also do this on Linux. Nevertheless, I'll modify the tests to avoid using `%` at all in the macros on the command line.

PiotrZSL commented on clang-tools-extra/clang-tidy/utils/FormatStringConverter.cpp:
```diff
@@ -230,6 +241,37 @@ FormatStringConverter::FormatStringConverter(ASTContext *ContextIn,
+      if (!MacroName.starts_with("PRI") && !MacroName.starts_with("__PRI"))
+        return MacroName;
+
+      const SourceLocation TokenSpellingLoc = SM.getSpellingLoc(TokenLoc);
```
add checking here if location is valid.

mikecrowe replied:
Do you mean that I should check whether the `SourceLocation` returned from `SourceManager::getSpellingLoc()` is valid?

Looking at the `SourceLocation` documentation and other checks that call `SourceManager::getSpellingLoc()` I was unable to spot how to tell whether the location returned is valid. All the other callers I looked at just went on to use the returned location without any checking.

Can I arrange for an invalid location to be returned from the lit test somehow to test this case?

PiotrZSL reviewed (approved):
Overall looks fine for me.

mikecrowe commented:
Unfortunately, walking the format string looking for macros also seems to match a macro that encloses the whole function invocation. This results in the check failing to work on expressions inside Catch2 `REQUIRE` macros for example.

My new test case (that currently fails) is:
```c++
#define SURROUND(x) x
  SURROUND(printf("Surrounding macro %d\n", 42));
  // CHECK-MESSAGES: [[@LINE-1]]:3: warning: use 'std::println' instead of 'printf' [modernize-use-std-print]
  // CHECK-FIXES: std::print("Surrounding macro {}", 42);
```
I'll try to fix this, but any advice is gratefully received.

mikecrowe commented:
> Unfortunately, walking the format string looking for macros also seems to match a macro that encloses the whole function invocation. This results in the check failing to work on expressions inside Catch2 `REQUIRE` macros for example.
> 
> My new test case (that currently fails) is:
> 
> ```c++
> #define SURROUND(x) x
>   SURROUND(printf("Surrounding macro %d\n", 42));
>   // CHECK-MESSAGES: [[@LINE-1]]:3: warning: use 'std::println' instead of 'printf' [modernize-use-std-print]
>   // CHECK-FIXES: std::print("Surrounding macro {}", 42);
> ```
> 
> I'll try to fix this, but any advice is gratefully received.

I've pushed a new commit that works for the Lit test cases that I've added by getting the macro name for the function call and permitting that macro for the format string too. I can't say that I'm particularly happy with it, and there may be corner cases such as the same macro being intentionally used in both places, but it's the best I can currently work out how to do.

5chmidti commented on clang-tools-extra/clang-tidy/utils/FormatStringConverter.cpp:
```diff
@@ -230,6 +241,50 @@ FormatStringConverter::FormatStringConverter(ASTContext *ContextIn,
 }
 
+std::optional<StringRef>
+FormatStringConverter::formatStringContainsUnreplaceableMacro(
```
> getting the macro name for the function call and permitting that macro for the format string too

>  such as the same macro being intentionally used in both places

It should be possible to detect this case comparing the `FileID` of the call's location and the format expression's location. If they differ, then they are different arguments -> different macros. Although, I haven't checked that yet.

mikecrowe commented:
@5chmidti wrote:
> It should be possible to detect this case comparing the FileID of the call's location and the format expression's location. If they differ, then they are different arguments -> different macros. Although, I haven't checked that yet.

I had thought so too but I was unable to find any `SourceLocation` or `Lexer` functions that returned what I needed. I've pushed an extra commit to the branch for this PR that contains an extra LIT test and some debug output. The test contains:
```C++
SURROUND_ALL(printf(SURROUND_FORMAT_PARTIAL("%" FMT) "\n", "42"));
```
with the hope that I can detect that `SURROUND_ALL` is applied to both the `printf` and the `"\n"` but different macros are applied to the other tokens that make up the format string.

The debug output (with the filenames shortened for readability) is:
```
getImmediateExpansionRange for 'BeginCallLoc': <use-std-print-macro.cpp.tmp.cpp:16:3 <Spelling=line:10:25>>
getExpansionRange for 'BeginCallLoc': <use-std-print-macro.cpp.tmp.cpp:16:3, col:67>
getTopMacroCaller for 'BeginCallLoc': use-std-print-macro.cpp.tmp.cpp:16:16
getImmediateMacroCallerLoc for 'BeginCallLoc': use-std-print-macro.cpp.tmp.cpp:16:16
getImmediateSpellingLoc for 'BeginCallLoc': use-std-print-macro.cpp.tmp.cpp:16:16
getFileLoc for 'BeginCallLoc': use-std-print-macro.cpp.tmp.cpp:16:16

getImmediateExpansionRange for 'SURROUND_FORMAT_PARTIAL': <use-std-print-macro.cpp.tmp.cpp:16:3 <Spelling=line:10:25>>
getExpansionRange for 'SURROUND_FORMAT_PARTIAL': <use-std-print-macro.cpp.tmp.cpp:16:3, col:67>
getTopMacroCaller for 'SURROUND_FORMAT_PARTIAL': use-std-print-macro.cpp.tmp.cpp:16:47
getImmediateMacroCallerLoc for 'SURROUND_FORMAT_PARTIAL': use-std-print-macro.cpp.tmp.cpp:16:23 <Spelling=use-std-print-macro.cpp.tmp.cpp:16:47>
getImmediateSpellingLoc for 'SURROUND_FORMAT_PARTIAL': use-std-print-macro.cpp.tmp.cpp:16:23 <Spelling=use-std-print-macro.cpp.tmp.cpp:16:47>
getFileLoc for 'SURROUND_FORMAT_PARTIAL': use-std-print-macro.cpp.tmp.cpp:16:47

getImmediateExpansionRange for 'FMT': <use-std-print-macro.cpp.tmp.cpp:16:3 <Spelling=line:10:25>>
getExpansionRange for 'FMT': <use-std-print-macro.cpp.tmp.cpp:16:3, col:67>
getTopMacroCaller for 'FMT': use-std-print-macro.cpp.tmp.cpp:16:51 <Spelling=use-std-print-macro.cpp.tmp.cpp:12:13>
getImmediateMacroCallerLoc for 'FMT': use-std-print-macro.cpp.tmp.cpp:16:23 <Spelling=use-std-print-macro.cpp.tmp.cpp:12:13>
getImmediateSpellingLoc for 'FMT': use-std-print-macro.cpp.tmp.cpp:16:23 <Spelling=use-std-print-macro.cpp.tmp.cpp:12:13>
getFileLoc for 'FMT': use-std-print-macro.cpp.tmp.cpp:16:51

getImmediateExpansionRange for 'SURROUND_ALL': <use-std-print-macro.cpp.tmp.cpp:16:3 <Spelling=line:10:25>>
getExpansionRange for 'SURROUND_ALL': <use-std-print-macro.cpp.tmp.cpp:16:3, col:67>
getTopMacroCaller for 'SURROUND_ALL': use-std-print-macro.cpp.tmp.cpp:16:56
getImmediateMacroCallerLoc for 'SURROUND_ALL': use-std-print-macro.cpp.tmp.cpp:16:56
getImmediateSpellingLoc for 'SURROUND_ALL': use-std-print-macro.cpp.tmp.cpp:16:56
getFileLoc for 'SURROUND_ALL': use-std-print-macro.cpp.tmp.cpp:16:56
```
Although the last token inside `SURROUND_ALL` has a few functions which return the same location as for `BeginCallLoc`, these functions also return that location for the tokens inside other macros too. :(

The change also captures the FileID for `BeginCallLoc` and compares it against each `TokenLoc` but never gets a match.

Or perhaps I've misinterpreted what you meant. If so, please can you provide more details?

5chmidti reviewed (commented):
It'll be 1 week+ before I am back at my main machine to take a better look at the macro question. Maybe this can just be landed, without that distinction.

mikecrowe commented:
@5chmidti wrote:
> It'll be 1 week+ before I am back at my main machine to take a better look at the macro question. Maybe this can just be landed, without that distinction.

Thanks for letting me know and thanks for the review.

I've pushed a new version to this PR with the typo you spotted fixed and removing the commit containing the debug output for exploring the macro detection (the version with the extra debug is at https://github.com/mikecrowe/clang-tidy-fmt/commits/mac/fix-use-std-print-macro-expansion-investigation/ if you still need it).

I'll let you and @PiotrZSL decide whether you want to land this PR as it is since I don't have permission to do so myself.

5chmidti reviewed (approved):
Looks good from my side

mikecrowe commented:
@5chmidti wrote:
> Looks good from my side

Thanks. I've squashed and rebased. Since 19 has been branched now and it includes *modernize-use-std-format* I've mentioned the improvement as being for both *modernize-use-std-print* and *modernize-use-std-format* in the release notes whilst I was resolving the conflict.

mikecrowe commented:
I've fixed the release notes conflicts and slightly tweaked the documentation a little. @5chmidti or @PiotrZSL, please land this if you are happy that it's a net improvement on the existing code. I can then try to work out how to improve it further. Thanks.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -194,10 +194,20 @@ Changes in existing checks
   <clang-tidy/checks/modernize/use-std-format>` check to support replacing
   member function calls too.
 
+- Improved :doc:`modernize-use-std-format
```
Should be merged with previous entry.

mikecrowe committed:
Merge entries in release notes

5chmidti commented:
Yep, thanks. Sorry for the delay
